### PR TITLE
fix(engine): replay backlog to all certified sessions on node reconnect

### PR DESCRIPTION
**Pull request by AgentRelayBot**
## Summary

Deliveries queued while a node's delivery socket is down were never replayed on reconnect — observed live: a GitHub webhook event landed in an agent's channel while its broker socket was down, and after the socket recovered the event sat durable-but-unsent until a *new* event arrived.

Root cause: `inventory.sync` replay on a cursor-negotiated reconnect was scoped to newly-routed agents only, so deliveries queued for already-ready agents had nothing to re-mark them due. This change makes reconnect replay every session the `inventory.sync` certifies, with the replay scope taken from the connection's negotiated delivery mode — connection-scoped readiness (`providerDeliveryReadinessMode()`) rather than `node_providers.capabilities`, which `heartbeatNode` rewrites wholesale and could silently un-flush the backlog.

- Legacy sessions keep register-time replay without a redundant inventory flush
- Registered cursor capability is preserved across heartbeat roster refreshes
- Compatibility fallback retained for older/out-of-process registry contracts

#### Test plan

- [x] 3 new conformance tests — all verified to fail against pre-fix code:
  - replay backlog when heartbeat roster omits cursor capability
  - no re-flush when heartbeat adds cursor capability to an immediate connection
  - registered cursor advertisement kept out of persisted heartbeat capability updates
- [x] `npx vitest run` in `packages/engine`: 99 files, 1,151 tests passed
- [x] Engine typecheck + lint clean
- [x] Full repo `npx turbo test`: 18/18 tasks

#### Known limits (documented, not introduced here)

- Transport remains at-least-once; the broker's monotonic-seq gate is the duplicate filter. This PR does not claim exactly-once.
- A half-open duplicate socket can mark a row `delivered` into a dead socket (replay still recovers it).
- Provider-less brokers synthesize a connection-based instance id, so reconnects can be classified as duplicate instances until the old incumbent goes stale.
- Cloud Durable Object socket-owner behavior not exercised directly.

Review evidence: `FLOW_NOTES.md` + `REVIEW_VERDICT.json` on this branch (fresh-eyes Codex review, approved at HEAD).

Generated with [Devin](https://devin.ai)

---

> [!NOTE]
> **Medium Risk**
> Changes core node reconnect delivery replay and heartbeat capability merging; behavior is well covered by new conformance tests but affects at-least-once message delivery to brokers.
> 
> **Overview**
> Fixes **stranded mailbox rows** when a broker node reconnects after its delivery socket was down: messages queued during the outage were not replayed on `inventory.sync` if the socket owner already reported those agents as delivery-ready.
> 
> **`inventory.sync` replay scope** now follows the connection’s negotiated delivery mode instead of readiness/routing *transitions*. On **cursor-negotiated** reconnects (no register-time flush), replay covers every identity in the certified inventory (`reconciledAgentIds`). **Legacy immediate-delivery** connections still only replay identities newly routed on that sync, avoiding a duplicate flush after `node.register`.
> 
> **Handshake stability:** replay mode is read from the live connection via optional `NodeConnectionRegistry.providerDeliveryReadinessMode()` (implemented on the in-process adapter). `node.heartbeat` roster updates merge capabilities with `withRegisteredProtocolCapabilities()` so heartbeats cannot drop or add `relay:delivery-cursor-v1` and silently change replay behavior. Superseded connections get `undefined` from the registry and do not drain backlog on stale syncs.
> 
> Conformance tests cover outage replay, ordering/dedupe after ACK, heartbeat roster edge cases, and stale-connection certification. Root and engine changelogs add an `[Unreleased - Patch]` fix entry; review notes and trajectory artifacts are included in the branch.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c8c70fc3e474e05da21aba676cebd4b12144a875. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

<a href="https://cubic.dev/pr/AgentWorkforce/relaycast/pull/445?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: packages/engine/src/engine/node.ts [hunk 1]
```diff
@@ -201,6 +201,84 @@ function supportsProviderDeliveryReadiness(registry: NodeConnectionRegistry): bo
...
+  registry: NodeConnectionRegistry,
+  workspaceId: string,
+  nodeId: string,
+  providerName: string,
+  connectionId: string | undefined,
+): Promise<boolean> {
+  if (!supportsProviderDeliveryReadiness(registry)) return false;
+  const mode = registry.providerDeliveryReadinessMode?.(workspaceId, nodeId, providerName, connectionId);
+  if (mode !== undefined) return mode === 'agent_scoped';
+  return providerAdvertisesDeliveryCursor(db, workspaceId, nodeId, providerName);
```
[the remark this answers is not in the document]
Comment (AgentRelayBot) [reply packages/engine/src/engine/node.ts]: Addressed in c8c70fc3: when providerDeliveryReadinessMode exists, an undefined result (stale/superseded connection) now means not-cursor-gated — the persisted capabilities fallback is only used when the API itself is absent. New conformance test 'does not let a stale connection certify a replacement connection''s backlog' covers exactly this path and fails without the fix.
